src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/IRequestHandler.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+#nullable enable
+
+namespace CommonLanguageServerProtocol.Framework;
+
+/// <summary>
+/// Top level type for LSP request handler.
+/// </summary>
+public interface IRequestHandler<RequestContextType>
+{
+    /// <summary>
+    /// Whether or not the solution state on the server is modified
+    /// as a part of handling this request.
+    /// </summary>
+    bool MutatesSolutionState { get; }
+
+    /// <summary>
+    /// Whether or not the handler execution queue should build a solution that represents the LSP
+    /// state of the world. If this property is not set <see cref="RequestContext.Solution"/> will be <see langword="null"/>
+    /// and <see cref="RequestContext.Document"/> will be <see langword="null"/>, even if <see cref="IRequestHandler{RequestType, ResponseType, RequestContextType}.GetTextDocumentIdentifier(RequestType)"/>
+    /// doesn't return null. Handlers should still provide text document information if possible to
+    /// ensure the correct workspace is found and validated.
+    /// </summary>
+    bool RequiresLSPSolution { get; }
+}
+
+public interface IRequestHandler<RequestType, ResponseType, RequestContextType> : IRequestHandler<RequestContextType>
+{
+    /// <summary>
+    /// Gets the <see cref="TextDocumentIdentifier"/> from the request, if the request provides one.
+    /// </summary>
+    TextDocumentIdentifier? GetTextDocumentIdentifier(RequestType request);
```

Comment:
Was hoping there was a better way to extract the document information from the request. After looking through the LSP spec there's TextDocumentPositionParams but not ever request uses those as a base type 😢. Maybe we can abstract away this detail in the future

Reply:
Yeah, I agree it's a bit cumbersome but I lacked for ideas on how to avoid it. Something to consider fixing before we 1.0 for sure.

Reply:
I guess one problem is that inspecting it for `TextDocumentPositionParams` is forcing everyone to use the VS contracts for their RequestParams types. We're already tied into that a little bit because we return TextDocumentIdentifier here but if anything I'd be more inclined to walk some of our commitments to VSContracts in the Framework itself back rather than spread them more.

Reply:
Perhaps there isn't any more, but I believe there was also one case (CodeActionResolve maybe?) where the document info was stored in a Data object of the json request even, and uses JTokens to find it